File: examples/jmediator-example-handlers/src/main/java/jmediator/example/handlers/ExampleCommandHandler.java
package jmediator.example.handlers;

import jmediator.example.ExampleCommand;
import jmediator.example.ExampleCommandResponse;
import jmediator.example.ExampleQuery;
import jmediator.example.domain.model.ExampleEntity;
import jmediator.example.domain.model.ExampleRepository;
import jmediator.MessageBus;
import jmediator.MessageHandler;
import reactor.core.publisher.Mono;

public class ExampleCommandHandler implements MessageHandler<ExampleCommand, ExampleCommandResponse> {
  private final MessageBus bus;
  private final ExampleRepository repository;

  public ExampleCommandHandler(
    final MessageBus bus,
    final ExampleRepository repository) {
    this.bus = bus;
    this.repository = repository;
  }

  @Override
  public Mono<ExampleCommandResponse> handle(ExampleCommand message) {
    var entity = repository.save(new ExampleEntity());
    return bus
      .send(new ExampleQuery() {{
        setFilter("some filter: " + entity.getId());
      }}).map(x -> new ExampleCommandResponse() {{
        setEchoDate(message.getDate());
      }});
  }
}

File: examples/jmediator-example-handlers-kotlin/src/main/kotlin/jmediator/example/handlers/ExampleCommandHandler.kt
package jmediator.example.handlers

import jmediator.example.ExampleCommand
import jmediator.example.ExampleCommandResponse
import jmediator.example.ExampleQuery
import jmediator.example.domain.model.ExampleEntity
import jmediator.example.domain.model.ExampleRepository
import jmediator.MessageBus
import jmediator.MessageHandler
import reactor.core.publisher.Mono

class ExampleCommandHandler(
  private val bus: MessageBus,
  private val repository: ExampleRepository
) : MessageHandler<ExampleCommand, ExampleCommandResponse> {

  override fun handle(message: ExampleCommand): Mono<ExampleCommandResponse> {

    val entity = repository.save(ExampleEntity())

    return bus
      .send(ExampleQuery().apply { filter = "(kotlin)Some filter: ${entity.id}" })
      .map { ExampleCommandResponse().apply { echoDate = message.date } }
  }
}